tests/test_base.py

```diff
@@ -336,3 +336,211 @@ def test_ensure_text_inputs_errors():
+    for i in range(1, 10):
+        assert grader(None, 'dog', attempt=i) == expected_result
+
+    # Ensure that zero credit is graded as false appropriately
+    grader = StringGrader(
+        answers='cat',
+        attempt_based_credit=True,
+        minimum_credit=0,
+        attempt_based_credit_msg=False
+    )
+
+    expected_result = {'msg': '', 'grade_decimal': 0, 'ok': False}
+    for i in range(10, 20):
+        assert grader(None, 'cat', attempt=i) == expected_result
+
+    # Ensure that messages are included as appropriate
+    grader = StringGrader(
+        answers='cat',
+        attempt_based_credit=True,
+        minimum_credit=0,
+        decrease_credit_after=1,
+        decrease_credit_steps=2,
+        attempt_based_credit_msg=True
+    )
+
+    expected_result = {'msg': '', 'grade_decimal': 1, 'ok': True}
+    assert grader(None, 'cat', attempt=1) == expected_result
+    expected_result = {'msg': '', 'grade_decimal': 0, 'ok': False}
+    assert grader(None, 'dog', attempt=1) == expected_result
+
+    expected_result = {'msg': 'Maximum credit for attempt #2 is 0.50.', 'grade_decimal': 0.5, 'ok': 'partial'}
+    assert grader(None, 'cat', attempt=2) == expected_result
+    expected_result = {'msg': '', 'grade_decimal': 0, 'ok': False}
+    assert grader(None, 'dog', attempt=2) == expected_result
+
+    expected_result = {'msg': 'Maximum credit for attempt #3 is 0.00.', 'grade_decimal': 0, 'ok': False}
+    assert grader(None, 'cat', attempt=3) == expected_result
+    expected_result = {'msg': '', 'grade_decimal': 0, 'ok': False}
+    assert grader(None, 'dog', attempt=3) == expected_result
+
+    # Ensure that messages are appended as appropriate
+    grader = StringGrader(
+        answers={'expect': 'cat', 'msg': 'Meow!'},
+        attempt_based_credit=True,
+        minimum_credit=0,
+        decrease_credit_after=1,
+        decrease_credit_steps=2,
+        attempt_based_credit_msg=True,
+        wrong_msg='too bad'
+    )
+
+    expected_result = {'msg': 'Meow!', 'grade_decimal': 1, 'ok': True}
+    assert grader(None, 'cat', attempt=1) == expected_result
+    expected_result = {'msg': 'too bad', 'grade_decimal': 0, 'ok': False}
+    assert grader(None, 'dog', attempt=1) == expected_result
+
+    expected_result = {'msg': 'Meow!<br/>\n<br/>\nMaximum credit for attempt #2 is 0.50.', 'grade_decimal': 0.5, 'ok': 'partial'}
+    assert grader(None, 'cat', attempt=2) == expected_result
+    expected_result = {'msg': 'too bad', 'grade_decimal': 0, 'ok': False}
+    assert grader(None, 'dog', attempt=2) == expected_result
+
+    expected_result = {'msg': 'Meow!<br/>\n<br/>\nMaximum credit for attempt #3 is 0.00.', 'grade_decimal': 0, 'ok': False}
+    assert grader(None, 'cat', attempt=3) == expected_result
+    expected_result = {'msg': 'too bad', 'grade_decimal': 0, 'ok': False}
+    assert grader(None, 'dog', attempt=3) == expected_result
+
+    # Ensure that things behave when an attempt number is not passed in
```

Comment:
Is the behavior below really correct? Or should we raise a `ConfigError` to the effect of

"Attempt number not passed to grader as keyword argument `'attempt'`. The input tag attribute `cfn_extra_args="attempt"` may need to be set.

(This is a question, since I've never written an attempt-based problem.)

Reply:
Hmm. If the attempt number isn't passed in, I was using that as a mechanism for cheaply turning this feature off (useful when combined with the defaults). You're correct in that it's presently implemented as a "fail silently". Do you think its worth having a config error raised instead so that people know there's a second piece that they need to implement? Eh, seems reasonable to me. I'm going to do it.